handler/src/test/java/io/netty/handler/ssl/CloseNotifyTest.java

```diff
@@ -0,0 +1,235 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.UnpooledByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.ssl.util.SelfSignedCertificate;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.Collection;
+import java.util.Queue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLSession;
+
+import static io.netty.buffer.ByteBufUtil.writeAscii;
+import static io.netty.buffer.Unpooled.EMPTY_BUFFER;
+import static io.netty.handler.codec.ByteToMessageDecoder.MERGE_CUMULATOR;
+import static io.netty.handler.ssl.SslUtils.PROTOCOL_TLS_V1_2;
+import static io.netty.handler.ssl.SslUtils.PROTOCOL_TLS_V1_3;
+import static java.util.Arrays.asList;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+
+@RunWith(Parameterized.class)
+public class CloseNotifyTest {
+
+    private static final UnpooledByteBufAllocator ALLOC = UnpooledByteBufAllocator.DEFAULT;
+    private static final Object INACTIVE = new Object() {
+        @Override
+        public String toString() {
+            return "INACTIVE";
+        }
+    };
+
+    private final SslProvider provider;
+    private final String protocol;
+
+    public CloseNotifyTest(SslProvider provider, String protocol) {
+        this.provider = provider;
+        this.protocol = protocol;
+    }
+
+    @Parameterized.Parameters(name = "{index}: provider={0}, protocol={1}")
+    public static Collection<Object[]> data() {
+        return asList(new Object[][] {
+                { SslProvider.JDK, PROTOCOL_TLS_V1_2 },
+                { SslProvider.JDK, PROTOCOL_TLS_V1_3 },
+                { SslProvider.OPENSSL, PROTOCOL_TLS_V1_2 },
+                { SslProvider.OPENSSL, PROTOCOL_TLS_V1_3 },
+        });
+    }
+
+    @Test(timeout = 5000)
+    public void eventsOrder() throws Exception {
+        BlockingQueue<Object> clientEventQueue = new LinkedBlockingQueue<Object>();
```

Comment:
looks like there maybe an issue with CI configuration not linking OpenSSL. Other tests skip/ignore OpenSSL if not available (e.g. [OpenSslEngineTest.java](https://github.com/netty/netty/blob/4.1/handler/src/test/java/io/netty/handler/ssl/OpenSslEngineTest.java#L111)) as the tests should be able to run against JDK engine in isolation. If this test follows the same patterns (and you validate / run against OpenSSL locally) we can do a followup PR to fix the CI configuration issues so that OpenSSL links successfully and is included in CI test suites.

Reply:
The issue that this PR fixes is reproducible only with OpenSSL provider. I've added assume to skip it when OpenSSL is not available. It's ok that at least JDK11 runs all tests and I verified locally on JDK8.
After the CI is fixed, we should consider removing this "assume" statement.